### DISCUSSION TITLE
Waveshare Pi Pico 2" display

### DISCUSSION THREAD

**Discussion by MikeTheGent** [peterhinch/micropython-nano-gui#39] [Q&A]
Thanks for Nano GUI — it looks excellent. If I can get it working with my display I think it will be very useful.

I have a Waveshare 2” display I’d like to get working. Using the 1.44” configuration I get the aclock display in the centre of the screen. If I simply change the resolution to 320x240, I get garbage. I assume this means the display can be made to work, but there’s more to it than simply changing the resolution. Are you willing to give me some pointers?

Thanks, Mike

**Comment by peterhinch** [comment]
It's hard to support hardware I don't possess but I'll try. What display controller does it use? Which driver are you using to show the clock display?

**Comment by MikeTheGent** [comment]
I’ll post the details here tomorrow morning UK time.

Mike

Love one another. We don’t need more instructions; We need more examples.

Bob Goff

**Reply by MikeTheGent** [reply]
The display is one of these https://www.waveshare.com/wiki/Pico-LCD-2 with ST7789V controller. Appears to be very similar to their 1.44" display, bigger with more pixels. Resolution is 320x240. I'm using the st7789.st7789_4bit driver.

**Reply by peterhinch** [reply]
There are two basic approaches: trial and error or doing a detailed study of the Waveshare example code.

When developing display drivers I always use [this code](https://github.com/peterhinch/micropython-nano-gui#23-verifying-hardware-configuration) to check: if the result is pixel perfect as described in the code comments then `nano-gui` will run. The simple nature of the graphic can indicate what is wrong if it doesn't display correctly.

I assume you're running in landscape mode and setting the width to 320. Perhaps you might post an image of the outcome of the above test with those args.

**Reply by MikeTheGent** [reply]
![IMG_1356](https://user-images.githubusercontent.com/39413591/188266863-df965819-cfea-4da4-89ed-a30bb6153b68.png)
![IMG_1357](https://user-images.githubusercontent.com/39413591/188266864-831b9eed-0e76-42f8-8676-01b420e13c60.png)
![IMG_1358](https://user-images.githubusercontent.com/39413591/188266868-c6dc5274-3a7d-4ecd-8548-eabea8fc3bd0.png)

The first image is the unmodified code which is encouraging  -- we've got the smaller screen in the centre of the 240x320 screen. But it's in the centre, not top left which explains what follows:

In the second image I have changed the width to 320, left the height at 135.
In the third image I have changed the height to 240.

The problem seems to be the offset of the display.

**Reply by MikeTheGent** [reply]
Just noticed the pattern that doesn't show up in photo 3, here's a better image. These small displays are hard to photograph with smart phone!

![Moment-App-20220903134228832](https://user-images.githubusercontent.com/39413591/188270968-4d7ed8df-983d-420d-883e-303fc7f73843.png)

**Comment by peterhinch** [comment]
The first image has the correct orientation but is evidently offset in X and Y. I would start by trying to correct the offsets.

You'll need to play with the `display` constructor arg. To do this, comment out [this check](https://github.com/peterhinch/micropython-nano-gui/blob/97d3be6dc96183bc472b01c0c5f0a09cbd04b82c/drivers/st7789/st7789_4bit.py#L64). The display arg is a 3-tuple, the default (which I assume you're using) is `(0, 0, 0)`. The first two elements are X and Y offsets. Try passing small values and observe the effects. The third element should stay as 0 because the orientation is correct.

If you can get the image _exactly_ located at the top left then you may be able to adjust height and width.

**Reply by MikeTheGent** [reply]
Thanks. I will experiment today.

**Comment by MikeTheGent** [comment]
Success! Based set up on setup_examples/st7789_pico_lcd_114.py, changed height & width. Changed drivers/st7789/st7789_4bit.py TDISPLAY to (0, 0, 1)

It works with TDISPLAY orientation. Many thanks for your help with this.

![IMG_1362](https://user-images.githubusercontent.com/39413591/188304297-ccec7d4a-aae0-46cc-a196-fceb06df6184.png)
![IMG_1363](https://user-images.githubusercontent.com/39413591/188304302-cf547f01-aff4-4ee8-83e4-7e97c9c4191b.png)

**Comment by peterhinch** [comment]
Excellent!

I am updating the driver to add a `PI_PICO_LCD_2` tuple with value `(0, 0, 1)`. I will also change the code so that an unexpected value of `display` merely prints a warning. This facilitates experimentation for further display types.

Please could you post a copy of your `color_setup.py` so that I can update the docs to include this very useful display.

**Comment by MikeTheGent** [comment]
Setup file:

[color_setup.py.zip](https://github.com/peterhinch/micropython-nano-gui/files/9485126/color_setup.py.zip)

**Comment by peterhinch** [comment]
I've now pushed an update documenting this driver [here](https://github.com/peterhinch/micropython-nano-gui/blob/master/DRIVERS.md#333-waveshare-pico-lcd-2).

Thank you for your efforts in enabling me to support this useful (and remarkably inexpensive) display.

**Reply by MikeTheGent** [reply]
Thanks for letting me know. Pleased to have made a small contribution.

**Comment by JumpinJon** [comment]
Hi,

I have one of these displays bought from Pi Hut: https://thepihut.com/products/2-ips-lcd-display-for-raspberry-pi-pico
Initialising with PI_PICO_LCD_2 works fine, and I've got plenty of code working.

I note the Pi Hut page states Driver: ST7789VW and 65K RGB Colours; and indeed shows a thumbnail with a bright landscape image.
I wonder if you are able to add support for the additional capability?  What can I provide in the way of diagnostics?

TIA

**Reply by peterhinch** [reply]
The display is 320*240 pixels. To support 65536 colors implies two bytes per pixel giving a buffer size of 153,600 bytes; this is quite unfeasible on the Pico. My drivers reduce the color information to four bits per pixel to overcome this.

The C drivers that use this display on the Pico use a different technique where commands are sent to the display avoiding the need for a frame buffer.

Drivers for nanogui and microgui are all framebuffer-based. This enables drivers to be minimal and also enables the GUI code to be device-independent. The answer therefore is a categorical no. If you want to display full-color pitcures you will need to use drivers not based on farmebuffers.

**Reply by JumpinJon** [reply]
Thank you for the detailed reply, that makes perfect sense.
